lib/std/simd.zig

```diff
@@ -0,0 +1,395 @@
+//! This module provides functions for working conveniently with SIMD (Single Instruction; Multiple Data),
+//! which may offer a potential boost in performance on some targets by performing the same operations on
+//! multiple elements at once.
+//! Please be aware that some functions are known to not work on MIPS.
+
+const std = @import("std");
+const builtin = @import("builtin");
+
+pub const Vector = std.meta.Vector;
+
+pub fn suggestVectorSizeForCpu(comptime T: type, cpu: std.Target.Cpu) ?usize {
+    switch (cpu.arch) {
+        .x86_64 => {
+            if (T == bool and std.Target.x86.featureSetHas(.prefer_mask_registers)) return 64;
+
+            const vector_bit_size = blk: {
+                if (std.Target.x86.featureSetHas(.avx512f)) break :blk 512;
+                if (std.Target.x86.featureSetHas(.prefer_256_bit)) break :blk 256;
+                if (std.Target.x86.featureSetHas(.prefer_128_bit)) break :blk 128;
+                return null;
+            };
+            const element_bit_size = std.math.max(8, std.math.ceilPowerOfTwo(T, @bitSizeOf(T)));
+            return @divExact(vector_bit_size, element_bit_size);
+        },
+        else => @compileError("No vector sizes for this CPU architecture have yet been recommended"),
+    }
+}
+
+/// Suggests a target-dependant vector size for a given type, or null if scalars are recommended.
+/// Not yet implemented for every CPU architecture.
+pub fn suggestVectorSize(comptime T: type) ?usize {
+    return suggestVectorSizeForCpu(T, builtin.cpu);
+}
+
+fn vectorLength(comptime VectorType: type) comptime_int {
+    return switch (@typeInfo(VectorType)) {
+        .Vector => |info| info.len,
+        .Array => |info| info.len,
+        else => @compileError("Invalid type " ++ @typeName(VectorType)),
+    };
+}
+
+/// Returns the smallest type of an unsigned int capable of indexing any element within the given vector type.
+pub fn VectorIndex(comptime VectorType: type) type {
+    return std.math.IntFittingRange(0, vectorLength(VectorType) - 1);
+}
+
+/// Returns the smallest type of an unsigned int capable of holding the length of the given vector type.
+pub fn VectorCount(comptime VectorType: type) type {
+    return std.math.IntFittingRange(0, vectorLength(VectorType));
+}
+
+/// Returns a vector containing the integer sequence starting at 0.
+/// For example, countUp(i32, 8) will return a vector containing {0, 1, 2, 3, 4, 5, 6, 7}.
+pub fn countUp(comptime T: type, comptime len: usize) Vector(len, T) {
```

Comment:
Other options include 'iota', or maybe something like 'indices'?

Reply:
> I would probably call it 'shifting left' and 'shifting right'.
(Somehow can't reply to the right comment)
So, what'd you consider "left" and "right"?

Reply:
Arrays are usually noted from left to right, when you write `[a, b, c]` index 0 is written on the left side, and so it seems to me that 'shifting left' is the same as 'shifting up'. Another option would be to generalize the two and use a signed value for the shift, though i must admit that i always forget whether to use positive or negative values for a particular rotate.

Reply:
How about `getIndices` or `genIndices`?

Reply:
I'm unsure what'd be left or right because on little-endian platforms like x86, shifting elements up is a left-shift in terms of bits.